### PR TITLE
[cxx] wasm-specific fixes --casts, goto-around-init, marshaling, externC

### PR DESCRIPTION
**Pull request by jaykrell**
https://jenkins.mono-project.com/job/test-mono-pull-request-wasm/3570/parsed_console/log.html

```/mnt/jenkins/workspace/test-mono-pull-request-wasm/mono/utils/mono-threads-wasm.c:133:12: error: assigning to 'guint8 *' (aka 'unsigned char *') from incompatible type 'void *'
        *staddr = (void*)wasm_get_stack_base ();
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/mnt/jenkins/workspace/test-mono-pull-request-wasm/mono/utils/mono-threads-wasm.c:172:7: error: no matching function for call to 'monoeg_g_slist_find'
        if (!g_slist_find (jobs, cb))
             ^~~~~~~~~~~~
                     ^
/mnt/jenkins/workspace/test-mono-pull-request-wasm/mono/utils/mono-threads-wasm.c:173:10: error: no matching function for call to 'monoeg_g_slist_prepend'
                jobs = g_slist_prepend (jobs, cb);
                       ^~~~~~~~~~~~~~~
3 errors generated.
```

https://jenkins.mono-project.com/job/test-mono-pull-request-wasm/3586/parsed_console/log.html

```
/mnt/jenkins/workspace/test-mono-pull-request-wasm/mono/metadata/gc.c:312:2: error: cannot jump from this goto statement to its label
        goto_if_nok (error, unhandled_error);
        ^
/mnt/jenkins/workspace/test-mono-pull-request-wasm/mono/utils/mono-error-internals.h:118:35: note: expanded from macro 'goto_if_nok'
#define goto_if_nok(error, label) goto_if (!is_ok (error), label)
                                  ^
/mnt/jenkins/workspace/test-mono-pull-request-wasm/mono/utils/mono-error-internals.h:116:48: note: expanded from macro 'goto_if'
#define goto_if(expr, label)      do { if (expr) goto label; } while (0)
                                                 ^
/mnt/jenkins/workspace/test-mono-pull-request-wasm/mono/metadata/gc.c:325:11: note: jump bypasses variable initialization
        gpointer params[] = { NULL };
```

### CODE REVIEW

**Comment by jaykrell** [comment]
@monojenkins build OS x x64

**Comment by jaykrell** [comment]
@monojenkins build Linux x64

**Comment by jaykrell** [comment]
@monojenkins build Linux WebAssembly

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 FullAOT

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64 Interpreter

**Comment by jaykrell** [comment]
@monojenkins build OS X i386

**Comment by jaykrell** [comment]
@monojenkins build Windows x64

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build OS X i386

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64

**Comment by jaykrell** [comment]
@monojenkins build Windows i386

**Comment by jaykrell** [comment]
Review?

File: mono/mini/mini-wasm.h [hunk 1]
```diff
@@ -54,7 +54,9 @@ typedef struct {
 
 void mono_wasm_debugger_init (void);
 
+G_BEGIN_DECLS
```
Comment (jaykrell) [inline mono/mini/mini-wasm.h]: wasm needs more externC than this.

Comment (luhenry) [inline mono/mini/mini-wasm.h]: Remove that. Use `MONO_API` if required

**Comment by jaykrell** [comment]
This  PR will probably be totally redundant soon. I'll keep it open just a bit longer.

**Comment by jaykrell** [comment]
Master should have all of this already.

**Pull request closed without merging**